src/Covenant/Type.hs

```diff
@@ -0,0 +1,443 @@
+-- much any case imaginable, this would be either 'AbstractTy' (in the ASG) or
+-- 'Renamed' (after renaming).
+--
+-- @ since 1.0.0
+data ValT (a :: Type)
+  = -- | An abstract type.
+    Abstraction a
+  | -- | A suspended computation.
+    ThunkT (CompT a)
+  | -- | A builtin type without any nesting.
+    BuiltinFlat BuiltinFlatT
+  | -- | A builtin type which is \'nested\'; something like lists, pairs etc.
+    BuiltinNested (BuiltinNestedT a)
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- Note (Koz, 04/03/2025): We use this for testing to compare for structural
+-- similarity.
+instance Eq1 ValT where
+  {-# INLINEABLE liftEq #-}
+  liftEq f = \case
+    Abstraction abs1 -> \case
+      Abstraction abs2 -> f abs1 abs2
+      _ -> False
+    ThunkT t1 -> \case
+      ThunkT t2 -> liftEq f t1 t2
+      _ -> False
+    BuiltinFlat t1 -> \case
+      BuiltinFlat t2 -> t1 == t2
+      _ -> False
+    BuiltinNested t1 -> \case
+      BuiltinNested t2 -> liftEq f t1 t2
+      _ -> False
+
+-- | All builtin types that are \'flat\': that is, do not have other types
+-- \'nested inside them\'.
+data BuiltinFlatT
+  = UnitT
+  | BoolT
+  | IntegerT
+  | StringT
+  | ByteStringT
+  | BLS12_381_G1_ElementT
+  | BLS12_381_G2_ElementT
+  | BLS12_381_MlResultT
+  | DataT
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- | Builtin types which have \'nested\' types. This is currently lists and
+-- pairs only.
+--
+-- = Important note
+--
+-- Both \'arms\' of this type have \'type abstraction boundaries\' just before
+-- them: their first field indicates how many type variables they bind. Note
+-- that 'PairT' has /one/ such boundary for both of its types, rather than one
+-- boundary per type.
+--
+-- While in truth, these types aren't /really/ polymorphic (as they cannot hold
+-- thunks, for example), we define them this way for now.
+--
+-- @since 1.0.0
+data BuiltinNestedT (a :: Type)
+  = ListT (Count "tyvar") (ValT a)
+  | PairT (Count "tyvar") (ValT a) (ValT a)
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- Note (Koz, 04/03/2025): We use this for testing to compare for structural
+-- similarity.
+instance Eq1 BuiltinNestedT where
+  {-# INLINEABLE liftEq #-}
+  liftEq f = \case
+    ListT abses1 t1 -> \case
+      ListT abses2 t2 -> abses1 == abses2 && liftEq f t1 t2
+      _ -> False
+    PairT abses1 t11 t12 -> \case
+      PairT abses2 t21 t22 -> abses1 == abses2 && liftEq f t11 t21 && liftEq f t12 t22
+      _ -> False
+
+-- Used during renaming. Contains a source of fresh indices for wildcards, as
+-- well as tracking:
+--
+-- 1. How many variables are bound by each scope;
+-- 2. Which of these variables have been noted as used; and
+-- 3. A unique identifier for each scope (for wildcards).
+data RenameState = RenameState Word64 (Vector (Vector Bool, Word64))
```

Comment:
I'm still a bit confused with the first `Word64`. What does it represent - it's not described in the comment IMO? I think it represents scopes count. If so, isn't that just an aggregation variable i.e. redundant information since the length of the vector is the same number?

Also, the `idSource` also doesn't say me much :thinking: 
Additionally, I don't think it's used somewhere, it's only written to. So maybe it's not even necessary?

Reply:
Also, the comment needs further update explaining the fields precisely:
First argument represents ...
Second represents tracker of each scope...
For an individual scope tracker `(Vector Bool, Word64)` the first argument is vector of all arguments and whether they've been used. The last argument is unique identifier of the scope...